### PR TITLE
KAFKA-18199: Fix size calculation for nullable tagged structs

### PR DESCRIPTION
**Pull request by squah-confluent**
When a struct field is tagged and nullable, it is serialized as
{ varint tag; varint dataLength; nullable<struct> data }, where
nullable<struct> is serialized as
{ varint isNotNull; if (isNotNull) struct s; }. The length field
includes the is-not-null varint.

This patch fixes a bug in serialization where the written value of
the length field and the value used to compute the size of the length
field differs by 1. In practice this has no impact unless the
serialized length of the struct is 127 bytes, since the varint encodings
of 127 and 128 have different lengths (0x7f vs 0x80 01).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/message/StructMessageTest.java [hunk 1]
```diff
@@ -27,46 +28,49 @@
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
-public class NullableStructMessageTest {
+public class StructMessageTest {
```
Comment (squah-confluent) [inline clients/src/test/java/org/apache/kafka/common/message/StructMessageTest.java]: Renamed `NullableStructMessageTest` to `StructMessageTest` since a non-nullable struct field is added.

File: generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java [hunk 2]
```diff
@@ -1343,6 +1343,11 @@ private void generateVariableLengthFieldSize(FieldSpec field,
                         buffer.printf("int _sizeBeforeStruct = _size.totalSize();%n", field.camelCaseName());
                         buffer.printf("this.%s.addSize(_size, _cache, _version);%n", field.camelCaseName());
                         buffer.printf("int _structSize = _size.totalSize() - _sizeBeforeStruct;%n", field.camelCaseName());
+                        VersionConditional.forVersions(field.nullableVersions(), possibleVersions).
+                            ifMember(__ -> {
+                                buffer.printf("_structSize += 1;%n");
+                            }).
+                            generate(buffer);
```
Comment (dajac) [inline generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java]: Don't we need to also change the code at L1336? I think that this is where the mistake is because we add one byte in all cases for nullable structs. It should be moved within the tagged/not-tagged branches in order to be correct. You did it for the tagged case. I suppose that we need to do the same for the not-tagged case.

I also wonder whether we should do it as follow:

```
buffer.printf("int _sizeBeforeStruct = _size.totalSize();%n", field.camelCaseName());
VersionConditional.forVersions(field.nullableVersions(), possibleVersions).
    ifMember(__ -> {
        buffer.printf("_size.addBytes(1);%n");
    }).
    generate(buffer);
buffer.printf("this.%s.addSize(_size, _cache, _version);%n", field.camelCaseName());
buffer.printf("int _structSize = _size.totalSize() - _sizeBeforeStruct;%n", field.camelCaseName());
buffer.printf("_size.addBytes(ByteUtils.sizeOfUnsignedVarint(_structSize));%n");
```

What do you think?

File: clients/src/test/resources/common/message/StructMessage.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 {
-  "name": "NullableStructMessage",
+  "name": "StructMessage",
```
Comment (dajac) [inline clients/src/test/resources/common/message/StructMessage.json]: Is this change needed? We do have other messages which contains non nullable structs. Could we extend those?

**Review by dajac — COMMENTED** [review]
@squah-confluent Nice find! I left a few comments.

File: generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java (hunk 2 above)
Comment (squah-confluent) [reply generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java]: I think I understand what you mean. The `+ 1` is due to the block at L1336 so we should move it into the size calculation below and clone it for the non-tagged case.

File: clients/src/test/resources/common/message/StructMessage.json (hunk 3 above)
Comment (squah-confluent) [reply clients/src/test/resources/common/message/StructMessage.json]: Now that I've addressed the other review comment, the rename no longer makes sense. 

`SimpleExampleMessageTest` and `NullableMessageTest` exercise all combinations of {non-tagged, tagged}x{non-nullable, nullable} structs, so adding a size assert to both those classes is simpler.
That just leaves the particular 127-long tagged nullable struct edge case. I've left the test for that in `NullableMessageTest`.

**Comment by squah-confluent** [comment]
Thanks for the review! I've tried to address the comments.

**Review by dajac — APPROVED** [review]
lgtm, thanks.

**Comment by dajac** [comment]
Merged to trunk and to 3.9.
